collector/hwmon_linux.go

```diff
@@ -0,0 +1,386 @@
+		"intrusion",
+	}
+)
+
+func init() {
+	Factories["hwmon"] = NewHwMonCollector
+}
+
+type hwMonCollector struct{}
+
+// Takes a prometheus registry and returns a new Collector exposing
+// /sys/class/hwmon stats (similar to lm-sensors).
+func NewHwMonCollector() (Collector, error) {
+	return &hwMonCollector{}, nil
+}
+
+func cleanMetricName(name string) string {
+	lower := strings.ToLower(name)
+	replaced := hwmonInvalidMetricChars.ReplaceAllLiteralString(lower, "_")
+	cleaned := strings.Trim(replaced, "_")
+	return cleaned
+}
+
+func addValueFile(data map[string]map[string]string, sensor string, prop string, file string) {
+	raw, e := ioutil.ReadFile(file)
+	if e != nil {
+		return
+	}
+	value := strings.Trim(string(raw), "\n")
+
+	if _, ok := data[sensor]; !ok {
+		data[sensor] = make(map[string]string)
+	}
+
+	data[sensor][prop] = value
+}
+
+// Split a sensor name into <type><num>_<property>
+func explodeSensorFilename(filename string) (ok bool, sensorType string, sensorNum int, sensorProperty string) {
+	matches := hwmonFilenameFormat.FindStringSubmatch(filename)
+	if len(matches) == 0 {
+		return false, sensorType, sensorNum, sensorProperty
+	}
+	for i, match := range hwmonFilenameFormat.SubexpNames() {
+		if i >= len(matches) {
+			return true, sensorType, sensorNum, sensorProperty
+		}
+		if match == "type" {
+			sensorType = matches[i]
+		}
+		if match == "property" {
+			sensorProperty = matches[i]
+		}
+		if match == "id" && len(matches[i]) > 0 {
+			if num, err := strconv.Atoi(matches[i]); err == nil {
+				sensorNum = num
+			} else {
+				return false, sensorType, sensorNum, sensorProperty
+			}
+		}
+	}
+	return true, sensorType, sensorNum, sensorProperty
+}
+
+func collectSensorData(dir string, data map[string]map[string]string) (err error) {
+	sensorFiles, dirError := ioutil.ReadDir(dir)
+	if dirError != nil {
+		return dirError
+	}
+	for _, file := range sensorFiles {
+		filename := file.Name()
+		ok, sensorType, sensorNum, sensorProperty := explodeSensorFilename(filename)
+		if !ok {
+			continue
+		}
+
+		for _, t := range hwmonSensorTypes {
+			if t == sensorType {
+				addValueFile(data, sensorType+strconv.Itoa(sensorNum), sensorProperty, path.Join(dir, file.Name()))
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func (c *hwMonCollector) updateHwmon(ch chan<- prometheus.Metric, dir string) (err error) {
+	hwmonName, err := c.hwmonName(dir)
+	if err != nil {
+		return err
+	}
+
+	data := make(map[string]map[string]string)
+	err = collectSensorData(dir, data)
+	if err != nil {
+		return err
+	}
+	if _, err := os.Stat(path.Join(dir, "device")); err == nil {
+		err := collectSensorData(path.Join(dir, "device"), data)
+		if err != nil {
+			return err
+		}
+	}
+
+	// format all sensors
+	for sensor, sensorData := range data {
+		label := sensor
+		if labelText, ok := sensorData["label"]; ok {
+			label = cleanMetricName(labelText)
+			if label == "" {
+				label = sensor
+			}
+		}
+		labels := []string{hwmonName, sensor, label}
```

Comment:
I'll give it a try on a few machines, if it works with just label I'd change to that. In theory it could fail with "bad" drivers, but I'm not sure if such drivers exist (sensor is guaranteed to be unique, label not).